src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java

```diff
@@ -390,6 +388,146 @@ private void resolveLocalCopies() {
+                p.getLocalCopy().getPath() + "\""));
+
+        // Turn it into an env var
+        p.setLocalCopy(null);
+        p.setType(null);
+                
+        // Move output folders to input parameters
+        // because outputs can only be files
+        if (!inputs.contains(p.getName())) {
+          defn.getOutputParameters().remove(p);
+          if (defn.getInputParameters() == null) {
+            defn.setInputParameters(new LinkedHashSet<Param>());
+          }
+          defn.getInputParameters().add(p);
+          
+          // move output folder args to input args
+          if (args.getOutputs() != null && args.getOutputs().containsKey(p.getName())) {
+            if (args.getInputs() == null) {
+              args.setInputs(new LinkedHashMap<String,String>());
+            }
+            args.getInputs().put(p.getName(), args.getOutputs().get(p.getName()));
+            args.getOutputs().remove(p.getName());
+          }
+        }
+      }
+    }
+    
+    // Pipelines API chokes on empty output parameter set
+    if (defn.getOutputParameters() != null && defn.getOutputParameters().isEmpty()) {
+      defn.setOutputParameters(null);
+    }
+    
+    // Edit the Docker command to copy the folders
+    StringBuilder copyInputs = new StringBuilder();
+    StringBuilder copyOutputs = new StringBuilder();
+    
+    for (String name : gcsPaths.keySet()) {
+      // Copy input folder
+      if (inputs.contains(name)) {
+        copyInputs.append(
+            "mkdir -p " +
+                DockerflowConstants.DEFAULT_MOUNT_POINT  + 
+                "/" +
+                localPaths.get(name) 
+                + " \n" +
+            "for ((i = 0; i < 3; i++)); do\n" +
+            "  if gsutil -m rsync -r "  + 
+                gcsPaths.get(name) + 
+                "/ " +
+                DockerflowConstants.DEFAULT_MOUNT_POINT + 
+                "/" +
+                localPaths.get(name) +
+                "/ ; then\n" +
+            "    break\n" +
+            "  elif ((i == 2)); then\n" +
+            "    2>&1 echo \"Recursive localization failed.\"\n" +
+            "    exit 1\n" +
+            "  fi\n" +
+            "done\n\n");
+      // Copy output folder
+      } else {
+        copyOutputs.append(
+            "for ((i = 0; i < 3; i++)); do\n" +
+            "  if gsutil -m rsync -r " + 
+                DockerflowConstants.DEFAULT_MOUNT_POINT +
+                "/" +
+                localPaths.get(name) + 
+                "/ " +
+                gcsPaths.get(name) + 
+                "/ ; then\n" +
+            "    break\n" +
+            "  elif ((i == 2)); then\n" +
+            "    2>&1 echo \"Recursive delocalization failed.\"\n" +
+            "    exit 1\n" +
+            "  fi\n" +
+            "done\n\n");
+      }
+    }
+    if (!gcsPaths.isEmpty()) {
+      defn.getDocker().setCmd(
+          "\n" +
+          "# Install gsutil\n" +
```

Comment:
I think this whole block of installation code should be pulled out to a separate string constant such that this block can be something more concise like:

    defn.getDocker().setCmd(
         "\n" +
         "# Install gsutil\n" +
         INSTALL_GCLOUD_CMD + "\n" +
         "# Copy inputs\n" +
         copyInputs.toString() + "\n" +
         "# Run user script\n" +
         defn.getDocker().getCmd() + "\n" +
         "# Copy outputs\n" +
         copyOutputs.toString());